src/tqec/compile/detectors/compute.py

```diff
@@ -0,0 +1,260 @@
+import stim
+
+from tqec.circuit.detectors.flow import build_flows_from_fragments
+from tqec.circuit.detectors.fragment import Fragment
+from tqec.circuit.detectors.match import (
+    MatchedDetector,
+    match_boundary_stabilizers,
+    match_detectors_within_fragment,
+)
+from tqec.circuit.generation import generate_circuit_from_instantiation
+from tqec.circuit.measurement import Measurement, get_measurements_from_circuit
+from tqec.circuit.qubit import GridQubit
+from tqec.circuit.schedule import ScheduledCircuit, relabel_circuits_qubit_indices
+from tqec.compile.detectors.database import DetectorDatabase
+from tqec.compile.detectors.detector import Detector
+from tqec.exceptions import TQECException
+from tqec.plaquette.plaquette import Plaquettes
+from tqec.position import Displacement, Position2D
+from tqec.templates.base import Template
+from tqec.templates.subtemplates import SubTemplateType
+
+
+def _get_measurement_offset_mapping(circuit: stim.Circuit) -> dict[int, Measurement]:
+    """Get a mapping from measurement offsets as used and returned by the
+    package `tqec.circuit.detectors` to a `Measurement` instance.
+
+    This function returns the mapping from negative offsets that are
+    supposed to each represent a unique measurement in the circuit to
+    `Measurement` instances, that serve the same purpose but use a different
+    encoding.
+
+    Note:
+        As a sanity check, the user of this function is encouraged to check
+        if the qubits in each `Measurement` instances returned as values of
+        the mapping corresponds to the expected qubit.
+
+    Args:
+        circuit: the circuit to create the mapping for.
+
+    Raises:
+        TQECException: if any qubit of the generated circuit is not a
+            `cirq.GridQubit` instance.
+
+    Returns:
+        a mapping qubit -> index for all the qubits appearing in the circuit
+        built from the provided `instantiation` and `plaquettes_at_timestep`.
+        Each qubit should be mapped to a different index.
+    """
+    return {
+        -i - 1: m
+        for i, m in enumerate(reversed(get_measurements_from_circuit(circuit)))
+    }
+
+
+def _matched_detectors_to_detectors(
+    detectors: list[MatchedDetector],
+    measurements_by_offset: dict[int, Measurement],
+    qubits_by_indices: dict[int, GridQubit],
+) -> frozenset[Detector]:
+    ret: list[Detector] = []
+    for d in detectors:
+        measurements: list[Measurement] = []
+        for m in d.measurements:
+            measurement = measurements_by_offset[m.offset]
+            expected_qubit = qubits_by_indices[m.qubit_index]
+            if measurement.qubit != expected_qubit:
+                raise TQECException(
+                    f"Expected qubit {expected_qubit} from index {m.qubit_index} "
+                    f"but got {measurement.qubit}."
+                )
```

Comment:
Is this check really needed?

Reply:
This is quite free to check (we should only have a few detectors here) and it actually already raised during the development due to mistakes I made in the code. So even though it is not strictly needed, it definitly helped and I think we may gain more leaving that code than removing it.